src/sgx/libpoet_enclave/poet_enclave.cpp

```diff
@@ -0,0 +1,690 @@
+            POET_LOG_ERROR,
+            "Unknown error in poet enclave(ecall_CreateSignupData)");
+        result = POET_ERR_UNKNOWN;
+    }
+
+    return result;
+} // ecall_CreateSignupData
+
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+poet_err_t ecall_VerifySignupInfo(
+    const sgx_target_info_t* inTargetInfo,
+    const char* inOriginatorPublicKeyHash,
+    const sgx_ec256_public_t* inPoetPublicKey,
+    sgx_report_t* outEnclaveReport
+    )
+{
+    poet_err_t result = POET_SUCCESS;
+
+    try {
+        sp::ThrowIfNull(
+            inTargetInfo,
+            "Target info pointer is NULL");
+        sp::ThrowIfNull(
+            inOriginatorPublicKeyHash,
+            "Originator public key hash pointer is NULL");
+        sp::ThrowIfNull(
+            inPoetPublicKey,
+            "PoET public key pointer is NULL");
+        sp::ThrowIfNull(
+            outEnclaveReport,
+            "Enclave report pointer is NULL");
+
+        // Create the report data we think should be given the OPK hash and the
+        // PPK.
+        sgx_report_data_t expectedReportData = { 0 };
+        CreateSignupReportData(
+            inOriginatorPublicKeyHash,
+            inPoetPublicKey,
+            &expectedReportData);
+
+        // Create the enclave report for the caller.
+        sgx_status_t ret =
+            sgx_create_report(
+                inTargetInfo,
+                &expectedReportData,
+                outEnclaveReport);
+        sp::ThrowSgxError(ret, "Failed to create enclave report");
+    } catch (sp::PoetError& e) {
+        Log(
+            POET_LOG_ERROR,
+            "Error in poet enclave(ecall_VerifySignupInfo): %04X -- %s",
+            e.error_code(),
+            e.what());
+        ocall_SetErrorMessage(e.what());
+        result = e.error_code();
+    } catch (...) {
+        Log(
+            POET_LOG_ERROR,
+            "Unknown error in poet enclave(ecall_VerifySignupInfo)");
+        result = POET_ERR_UNKNOWN;
+    }
+
+    return result;
+} // ecall_VerifySignupInfo
+
+uint64_t deserializeWaitCert(const char* waitCert) {
+    JsonValue parsed(json_parse_string(waitCert));
+    sp::ThrowIf<sp::ValueError>(!parsed.value, "Failed to parse SerializedWaitCertificate");
+    JSON_Object* pObject = json_value_get_object(parsed);
+    sp::ThrowIfNull(
+            pObject,
+            "Got Null JSON Object");
+
+    return json_object_dotget_number(pObject, "block_number");
+}
+
+void create_duration(uint8_t* duration, size_t inDurationLenBytes) {
+    sgx_status_t ret = sgx_read_rand(duration, inDurationLenBytes);
+    sp::ThrowSgxError(ret, "Failed to generate duration for wait certificate");
+}
+
+poet_err_t ecall_InitializeWaitCertificate(
+    const char* inPreviousWaitCertificate, 
+    size_t inPreviousWaitCertificateLen, 
+    const char* inValidatorId, 
+    size_t inValidatorIdLen,
+    const char* inPrevBlockId,
+    size_t inPrevBlockIdLen,
+    uint8_t* outDuration,
```

Comment:
The return for InitWC should be the duration truncated to 64 bits, i.e.u64.  Not the 256-bit duration itself.